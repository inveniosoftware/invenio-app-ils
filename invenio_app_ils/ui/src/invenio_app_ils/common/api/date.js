import { DateTime } from 'luxon';
import _get from 'lodash/get';

export const fromISO = stringDate => {
  return DateTime.fromISO(stringDate);
};

/**
 * Converts datetime JSON string to luxon Datetime object
 * @param date
 * @returns {DateTime}
 */
export const toISO = date => {
  return date.toISO();
};

/**
 *  Serializes date for display in short format
 *  @param {DateTime} date luxon DateTime
 *  @re
 */
export const toShortDateTime = date => {
  return date.toFormat('yyyy-MM-dd HH:mm');
};

/**
 *  Serializes date for display in short format
 *  @param {DateTime} date luxon DateTime
 *  @re
 */
export const toShortDate = date => {
  return date.toFormat('yyyy-MM-dd');
};

/**
 *  Serializes date for UTC date in short format used to
 *  normalise dates for REST API
 *  @param {DateTime} date luxon DateTime
 *  @re
 */
export const toUTCShortDate = date => {
  if (!(typeof date === 'string' || date instanceof DateTime)) {
    throw Error('Wrong date type provided');
  }
  if (typeof date === 'string') {
    date = fromISO(date);
  }
  return date.toUTC().toFormat('yyyy-MM-dd');
};

/**
 *  Serializes date for display in tables
 */
export function dateFormatter({ col, row }) {
  const dateField = _get(row, col.field);
  return dateField ? toShortDate(fromISO(dateField)) : null;
}
