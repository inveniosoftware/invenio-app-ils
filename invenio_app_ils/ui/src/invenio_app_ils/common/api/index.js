export { document } from './documents/document';
export { http } from './base';
export { item } from './items/item';
export { eitem } from './eitems/eitem';
export { loan } from './loans/loan';
export { patron } from './patrons/patron';
export { location } from './locations/location';
export { internalLocation } from './locations/internalLocation';
export { keyword } from './keywords/keyword';
export { series } from './series/series';
export { related } from './related/related';
