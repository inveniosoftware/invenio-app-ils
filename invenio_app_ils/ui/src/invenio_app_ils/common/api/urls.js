export const ApiURLS = {
  loans: {
    loan: '/circulation/loans/:loanPid',
    list: '/circulation/loans/',
    checkout: '/circulation/loans/checkout',
    request: '/circulation/loans/request',
  },
  items: {},
  documents: {},
  locations: {},
  users: {},
};
