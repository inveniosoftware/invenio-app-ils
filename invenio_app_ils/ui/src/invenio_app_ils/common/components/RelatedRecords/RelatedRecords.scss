.tabular.menu {
  margin-bottom: 0;
}

.related-records-form {
  .content {
    .header {
      span.extra {
        color: #21ba45;
        float: right;
        font-size: 0.85rem;
        font-weight: 500;
      }
    }
  }
}
