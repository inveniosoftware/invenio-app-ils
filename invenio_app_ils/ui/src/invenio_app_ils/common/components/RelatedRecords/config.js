import { invenioConfig } from '../../config';

export function getRelationTypes() {
  return invenioConfig.relationTypes;
}

export function getRelationTypeByName(name) {
  for (const type of getRelationTypes()) {
    if (type.name === name) {
      return type;
    }
  }
  throw Error(`No relation with name: ${name}`);
}

export function getRelationTypeById(id) {
  for (const type of getRelationTypes()) {
    if (type.id === id) {
      return type;
    }
  }
  throw Error(`No relation with id: ${id}`);
}

export function getIconByRelation(relation) {
  switch (relation.name) {
    case EditionRelation.name:
      return 'book';
    case LanguageRelation.name:
      return 'translate';
    default:
      return null;
  }
}

export const EditionRelation = getRelationTypeByName('edition');
export const LanguageRelation = getRelationTypeByName('language');
