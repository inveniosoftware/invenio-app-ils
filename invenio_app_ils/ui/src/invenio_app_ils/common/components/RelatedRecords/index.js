import { connect } from 'react-redux';

import { fetchRelatedRecords, updateRelatedRecords } from './state/actions';
import RelatedRecordsComponent from './RelatedRecords';

const mapStateToProps = state => ({
  data: state.relatedRecords.data,
  error: state.relatedRecords.error,
  isLoading: state.relatedRecords.isLoading,
  hasError: state.relatedRecords.hasError,
});

const mapDispatchToProps = dispatch => ({
  fetchRelatedRecords: (pid, pidType) =>
    dispatch(fetchRelatedRecords(pid, pidType)),
  updateRelatedRecords: (pid, pidType, data) =>
    dispatch(updateRelatedRecords(pid, pidType, data)),
});

export const RelatedRecords = connect(
  mapStateToProps,
  mapDispatchToProps
)(RelatedRecordsComponent);
