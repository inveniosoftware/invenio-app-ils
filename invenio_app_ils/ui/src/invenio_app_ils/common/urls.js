import { generatePath } from 'react-router-dom';
import { loan as loanApi } from './api/loan';
import { item as itemApi } from './api/item';

export const FrontSiteURLS = {};

const BackOfficeBase = '/backoffice';

export const BackOfficeURLS = {
  home: BackOfficeBase,
  documentDetails: `${BackOfficeBase}/documents/:documentPid`,
  itemsSearch: `${BackOfficeBase}/items`,
  itemDetails: `${BackOfficeBase}/items/:itemPid`,
  loansSearch: `${BackOfficeBase}/loans`,
  loanDetails: `${BackOfficeBase}/loans/:loanPid`,
  patronDetails: `${BackOfficeBase}/users/:userPid`,
};

export const viewLoanDetailsUrl = loanPid => {
  return generatePath(BackOfficeURLS.loanDetails, {
    loanPid: loanPid,
  });
};

export const showAllLoansUrl = (
  documentPid,
  itemPid,
  state,
  patronPid,
  extraQuery
) => {
  const qs = loanApi.buildLoansQuery(
    documentPid,
    itemPid,
    state,
    patronPid,
    extraQuery
  );
  return `${BackOfficeURLS.loansSearch}?q=${qs}`;
};

export const viewItemDetailsUrl = itemPid => {
  return generatePath(BackOfficeURLS.itemDetails, {
    itemPid: itemPid,
  });
};

export const showAllItemsUrl = (documentPid, state, extraQuery) => {
  const qs = itemApi.buildItemsQuery(documentPid, state, extraQuery);
  return `${BackOfficeURLS.itemsSearch}?q=${qs}`;
};
