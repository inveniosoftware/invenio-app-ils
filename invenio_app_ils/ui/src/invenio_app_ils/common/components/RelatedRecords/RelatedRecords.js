import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Loader, Error } from '../../components';
import { ResultsTable } from '../../components';
import { BackOfficeRoutes } from '../../../routes/urls';
import { SeeAllButton } from '../../../pages/backoffice/components/buttons';
import { goTo } from '../../../history';
import { formatter } from '../ResultsTable/formatters';
import pick from 'lodash/pick';
import truncate from 'lodash/truncate';
import { Grid, Button } from 'semantic-ui-react';
import { recordToPid } from '../../api/utils';
import ESRelatedSelector from '../ESSelector/ESRelatedSelector';

export default class RelatedRecords extends Component {
  constructor(props) {
    super(props);
    this.fetchRelatedRecords = props.fetchRelatedRecords;
    this.showDetailsUrl = BackOfficeRoutes.itemDetailsFor;
    this.seeAllUrl = BackOfficeRoutes.itemsListWithQuery;
    this.state = {
      removedRelatedRecords: [],
      showMaxRelatedEditions: props.showMaxRelatedRecords,
      showMaxRelatedTranslations: props.showMaxRelatedRecords,
    };
  }

  componentDidMount() {
    const [pid, pidType] = recordToPid(this.props.record);
    this.fetchRelatedRecords(pid, pidType);
  }

  onAllEditionsClick = () => {
    this.setState({ showMaxRelatedEditions: 10 });
  };

  onAllTranslationsClick = () => {
    this.setState({ showMaxRelatedTranslations: 10 });
  };

  seeAllButton = onClick => {
    return <SeeAllButton clickHandler={onClick} />;
  };

  onRelatedClick = row => {
    const pid = row.ID;
    const type = row.Type;
    let detailsFor;
    if (type === 'Document') {
      detailsFor = BackOfficeRoutes.documentDetailsFor;
    } else if (type === 'Series') {
      detailsFor = BackOfficeRoutes.seriesDetailsFor;
    } else {
      console.warn(`Unknown pid type: ${type}`);
    }
    return goTo(detailsFor(pid), { pid: pid, type: type });
  };

  onRemoveRelatedSelection = selection => {
    const removed = this.state.removedRelatedRecords;
    removed.push(selection);
    this.setState({ removedRelatedRecords: removed });
  };

  updateRelatedRecords = results => {
    const resultToAction = (result, action) => ({
      pid: result.metadata.pid,
      pid_type: result.metadata.pidType,
      relation_type: result.metadata.relationType,
      action: action,
    });
    const updatedRecords = [];
    for (const result of results) {
      if (result.metadata.new) {
        updatedRecords.push(resultToAction(result, 'add'));
      }
    }
    for (const result of this.state.removedRelatedRecords) {
      if (!result.metadata.new) {
        updatedRecords.push(resultToAction(result, 'remove'));
      }
    }
    const [pid, pidType] = recordToPid(this.props.record);
    this.props.updateRelatedRecords(pid, pidType, updatedRecords);
    this.setState({ removedRelatedRecords: [] });
  };

  prepareData(data) {
    const editions = [];
    const translations = [];
    const initialSelections = [];
    for (let obj of data.metadata.related_records) {
      const record = formatter.related.toTable(obj);
      if (record.Relation === 'Edition') {
        editions.push(pick(record, ['ID', 'Type', 'Title', 'Edition']));
      } else if (record.Relation === 'Translation') {
        translations.push(pick(record, ['ID', 'Type', 'Title', 'Language']));
      } else {
        console.warn(`Unknown record relation: ${record}`);
      }
      const id = `${obj.pid}-${obj.pid_type}-${obj.relation_type}`;
      initialSelections.push({
        id: id,
        key: id,
        title: truncate(obj.title, { length: 50 }),
        description: `${record.Type}`,
        extra: `PID: ${obj.pid}`,
        metadata: {
          pid: obj.pid,
          pidType: obj.pid_type,
          relationType: obj.relation_type,
        },
      });
    }
    return [editions, translations, initialSelections];
  }

  renderTable(data) {
    const [editions, translations, initialSelections] = this.prepareData(data);
    const seeAllEditions = this.seeAllButton(
      this.onAllEditionsClick,
      editions.length
    );
    const seeAllTranslations = this.seeAllButton(
      this.onAllTranslationsClick,
      translations.length
    );
    const { SelectorModal } = this.props;
    return (
      <Grid>
        <Grid.Row>
          <Grid.Column width={8}>
            <ResultsTable
              rows={editions}
              title={'Related editions'}
              name={'related editions'}
              rowActionClickHandler={this.onRelatedClick}
              seeAllComponent={seeAllEditions}
              showMaxRows={this.state.showMaxRelatedEditions}
            />
          </Grid.Column>
          <Grid.Column width={8}>
            <ResultsTable
              rows={translations}
              title={'Related translations'}
              name={'related translations'}
              rowActionClickHandler={this.onRelatedClick}
              seeAllComponent={seeAllTranslations}
              showMaxRows={this.state.showMaxRelatedTranslations}
            />
          </Grid.Column>
        </Grid.Row>
        <Grid.Row>
          <Grid.Column width={13} />
          <Grid.Column width={3}>
            {SelectorModal && (
              <SelectorModal
                multiple
                selectorComponent={ESRelatedSelector}
                initialSelections={initialSelections}
                trigger={<Button color="blue" content="Manage related" />}
                title="Select Related"
                size="small"
                content={
                  'Select related documents/series and their type of relation.'
                }
                onRemoveSelection={this.onRemoveRelatedSelection}
                onSave={this.updateRelatedRecords}
              />
            )}
          </Grid.Column>
        </Grid.Row>
      </Grid>
    );
  }

  render() {
    const { data, isLoading, error } = this.props;
    return (
      <Loader isLoading={isLoading}>
        <Error error={error}>{this.renderTable(data)}</Error>
      </Loader>
    );
  }
}

RelatedRecords.propTypes = {
  record: PropTypes.object.isRequired,
  fetchRelatedRecords: PropTypes.func.isRequired,
  data: PropTypes.object.isRequired,
  showMaxRelatedRecords: PropTypes.number,
  SelectorModal: PropTypes.func.isRequired,
};

RelatedRecords.defaultProps = {
  showMaxRelatedRecords: 5,
};
