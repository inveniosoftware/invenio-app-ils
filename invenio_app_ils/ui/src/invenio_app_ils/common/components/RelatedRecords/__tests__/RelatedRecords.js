import React from 'react';
import { shallow, mount } from 'enzyme';
import RelatedRecords from '../RelatedRecords';
import { Settings } from 'luxon';
import { fromISO } from '../../../api/date';
import history from '../../../../history';
import { BackOfficeRoutes } from '../../../../routes/urls';
import { Modal } from 'semantic-ui-react';

jest.mock('../../../config');

Settings.defaultZoneName = 'utc';
const stringDate = fromISO('2018-01-01T11:05:00+01:00');

describe('RelatedRecords tests', () => {
  let component;
  afterEach(() => {
    if (component) {
      component.unmount();
    }
  });

  const doc = {
    document_pid: '111',
    metadata: {
      pid: '111',
      $schema: 'https://127.0.0.1:5000/schemas/documents/document-v1.0.0.json',
      related_records: [],
    },
  };

  it('should load the RelatedRecords component', () => {
    const data = {
      metadata: {
        related_records: [],
      },
    };
    const component = shallow(
      <RelatedRecords
        record={doc}
        data={data}
        fetchRelatedRecords={() => {}}
        SelectorModal={Modal}
      />
    );
    expect(component).toMatchSnapshot();
  });

  it('should render show a message with no related records', () => {
    const data = {
      metadata: {
        related_records: [],
      },
    };
    component = mount(
      <RelatedRecords
        record={doc}
        data={data}
        fetchRelatedRecords={() => {}}
        SelectorModal={Modal}
      />
    );

    const message = component
      .find('Message')
      .filterWhere(element => element.prop('data-test') === 'no-results');
    expect(message).toHaveLength(2);
  });

  it('should render related record', () => {
    const data = {
      metadata: {
        related_records: [
          {
            pid: 'docid-1',
            pid_type: 'docid',
            relation_type: 0,
          },
        ],
      },
    };

    component = mount(
      <RelatedRecords
        record={doc}
        data={data}
        fetchRelatedRecords={() => {}}
        SelectorModal={Modal}
      />
    );

    const rows = component
      .find('TableRow')
      .filterWhere(element => element.prop('data-test') === 'docid-1');
    expect(rows).toHaveLength(1);

    const footer = component
      .find('TableRow')
      .filterWhere(element => element.prop('data-test') === 'footer');
    expect(footer).toHaveLength(0);
  });

  it('should render the see all button when showing only a few items', () => {
    const data = {
      metadata: {
        related_records: [
          {
            pid: 'docid-1',
            pid_type: 'docid',
            relation_type: 0,
          },
          {
            pid: 'docid-2',
            pid_type: 'docid',
            relation_type: 0,
          },
          {
            pid: 'docid-3',
            pid_type: 'docid',
            relation_type: 1,
          },
          {
            pid: 'docid-4',
            pid_type: 'docid',
            relation_type: 1,
          },
        ],
      },
    };

    component = mount(
      <RelatedRecords
        record={doc}
        data={data}
        fetchRelatedRecords={() => {}}
        showMaxRelatedRecords={1}
        SelectorModal={Modal}
      />
    );

    const footer = component
      .find('TableFooter')
      .first()
      .filterWhere(element => element.prop('data-test') === 'footer');
    expect(footer).toHaveLength(1);
  });

  it('should go to record details when clicking on a row', () => {
    const mockedHistoryPush = jest.fn();
    history.push = mockedHistoryPush;
    const data = {
      metadata: {
        related_records: [
          {
            pid: 'docid-1',
            pid_type: 'docid',
            relation_type: 0,
          },
          {
            pid: 'docid-2',
            pid_type: 'docid',
            relation_type: 1,
          },
        ],
      },
    };

    component = mount(
      <RelatedRecords
        record={doc}
        data={data}
        fetchRelatedRecords={() => {}}
        showMaxRelatedRecords={1}
        SelectorModal={Modal}
      />
    );

    const firstId = data.metadata.related_records[0].pid;
    const button = component
      .find('TableRow')
      .filterWhere(element => element.prop('data-test') === firstId)
      .find('i');
    button.simulate('click');

    const expectedUrl = BackOfficeRoutes.documentDetailsFor(firstId);
    const expectedState = { pid: 'docid-1', type: 'Document' };
    expect(mockedHistoryPush).toHaveBeenCalledWith(expectedUrl, expectedState);
  });
});
