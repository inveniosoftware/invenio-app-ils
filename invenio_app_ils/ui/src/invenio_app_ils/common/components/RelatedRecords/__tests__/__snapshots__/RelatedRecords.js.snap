// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RelatedRecords tests should load the RelatedRecords component 1`] = `
<Loader>
  <Error>
    <Grid>
      <GridRow>
        <GridColumn
          width={8}
        >
          <ResultsTable
            fixed={false}
            headerActionClickHandler={null}
            headerActionComponent={null}
            name="related editions"
            rowActionClickHandler={[Function]}
            rows={Array []}
            seeAllComponent={
              <SeeAllButton
                clickHandler={[Function]}
                disabled={false}
                fluid={false}
              />
            }
            showMaxRows={5}
            singleLine={false}
            subtitle=""
            title="Related editions"
          />
        </GridColumn>
        <GridColumn
          width={8}
        >
          <ResultsTable
            fixed={false}
            headerActionClickHandler={null}
            headerActionComponent={null}
            name="related translations"
            rowActionClickHandler={[Function]}
            rows={Array []}
            seeAllComponent={
              <SeeAllButton
                clickHandler={[Function]}
                disabled={false}
                fluid={false}
              />
            }
            showMaxRows={5}
            singleLine={false}
            subtitle=""
            title="Related translations"
          />
        </GridColumn>
      </GridRow>
      <GridRow>
        <GridColumn
          width={13}
        />
        <GridColumn
          width={3}
        >
          <Modal
            centered={true}
            closeOnDimmerClick={true}
            closeOnDocumentClick={false}
            content="Select related documents/series and their type of relation."
            dimmer={true}
            eventPool="Modal"
            initialSelections={Array []}
            multiple={true}
            onRemoveSelection={[Function]}
            onSave={[Function]}
            selectorComponent={[Function]}
            size="small"
            title="Select Related"
            trigger={
              <Button
                as="button"
                color="blue"
                content="Manage related"
              />
            }
          />
        </GridColumn>
      </GridRow>
    </Grid>
  </Error>
</Loader>
`;
